src/data-source/collection-browser-data-source.ts

```diff
@@ -0,0 +1,1189 @@
+   */
+  map(
+    callback: (model: TileModel, index: number, array: TileModel[]) => TileModel
+  ): void;
+
+  /**
+   * Checks every tile's management checkbox
+   */
+  checkAllTiles(): void;
+
+  /**
+   * Unchecks every tile's management checkbox
+   */
+  uncheckAllTiles(): void;
+
+  /**
+   * Removes all tile models that are currently checked & adjusts the paging
+   * of the data source to account for any new gaps in the data.
+   */
+  removeCheckedTiles(): void;
+}
+
+export class CollectionBrowserDataSource
+  implements CollectionBrowserDataSourceInterface
+{
+  private pages: Record<string, TileModel[]> = {};
+
+  private offset = 0;
+
+  private numTileModels = 0;
+
+  /**
+   * Maps the full query key to the pages being fetched for that query
+   */
+  private pageFetchesInProgress: Record<string, Set<number>> = {};
+
+  totalResults = 0;
+
+  endOfDataReached = false;
+
+  /**
+   * @inheritdoc
+   */
+  aggregations?: Record<string, Aggregation>;
+
+  /**
+   * @inheritdoc
+   */
+  yearHistogramAggregation?: Aggregation;
+
+  /**
+   * @inheritdoc
+   */
+  collectionTitles = new Map<string, string>();
+
+  /**
+   * @inheritdoc
+   */
+  collectionExtraInfo?: CollectionExtraInfo;
+
+  /**
+   * @inheritdoc
+   */
+  accountExtraInfo?: AccountExtraInfo;
+
+  /**
+   * @inheritdoc
+   */
+  pageElements?: PageElementMap;
+
+  /**
+   * @inheritdoc
+   */
+  parentCollections?: string[] = [];
+
+  /**
+   * @inheritdoc
+   */
+  prefixFilterCountMap: Partial<Record<PrefixFilterType, PrefixFilterCounts>> =
+    {};
+
+  constructor(
+    /** The host element to which this controller should attach listeners */
+    private readonly host: ReactiveControllerHost &
+      CollectionBrowserSearchInterface,
+    /** Default size of result pages */
+    private pageSize: number
+  ) {
+    this.host.addController(this as CollectionBrowserDataSourceInterface);
+  }
+
+  /**
+   * @inheritdoc
+   */
+  get size(): number {
+    return this.numTileModels;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  addPage(pageNum: number, pageTiles: TileModel[]): void {
+    this.pages[pageNum] = pageTiles;
+    this.numTileModels += pageTiles.length;
+    this.host.requestUpdate();
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getPage(pageNum: number): TileModel[] {
+    return this.pages[pageNum];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getAllPages(): Record<string, TileModel[]> {
+    return this.pages;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  hasPage(pageNum: number): boolean {
+    return !!this.pages[pageNum];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  getTileModelAt(index: number): TileModel | undefined {
+    const pageNum = Math.floor(index / this.pageSize) + 1;
+    const indexOnPage = index % this.pageSize;
+    return this.pages[pageNum]?.[indexOnPage];
+  }
+
+  /**
+   * @inheritdoc
+   */
+  setPageSize(pageSize: number): void {
+    this.reset();
+    this.pageSize = pageSize;
+  }
+
+  /**
+   * @inheritdoc
+   */
+  reset(): void {
```

Comment:
suggestion: raise this higher up the class so we know what a reset data source looks like